### PR TITLE
feat(slimfaas): call only at least when one pod is read

### PR DESCRIPTION
**Pull request by guillaume-chervet**
#16 

Now we wait at least one pod is in started state before doing the first HTTP try.

### CODE REVIEW

File: src/SlimFaas/Kubernetes/KubernetesService.cs [hunk 1]
```diff
@@ -62,7 +83,8 @@ public KubernetesService(IConfiguration config, ILogger<KubernetesService> logge
         }
         catch (HttpOperationException e)
         {
-            _logger.LogError(e, "Error while scaling kubernetes deployment" + request.Deployment);
+            var empty = "";
```
Comment (guillaume-chervet) [inline src/SlimFaas/Kubernetes/KubernetesService.cs:86]: a virere

File: src/SlimFaas/SlimProxyMiddleware.cs [hunk 2]
```diff
@@ -32,33 +33,79 @@ public async Task InvokeAsync(HttpContext context, ILogger<SlimProxyMiddleware>
...
+        if (function == null)
+        {
+            context.Response.StatusCode = 404;
+            return;
+        }
+
         historyHttpService.SetTickLastCall(functionName, DateTime.Now.Ticks);
+        var numerLoop = 100;
+        while (numerLoop > 0)
+        {
```
Comment (guillaume-chervet) [inline src/SlimFaas/SlimProxyMiddleware.cs:98]: Set this configurable

**Comment by guillaume-chervet** [comment]
Pour le merge je ferais un squash merge

File: src/SlimFaas/ReplicasService.cs [hunk 3]
```diff
@@ -110,11 +99,11 @@ public Task CheckScaleAsync(string kubeNamespace)
         if (tasks.Count <= 0) return Task.CompletedTask;
 
         var updatedFunctions = new List<DeploymentInformation>();
-        
+
         foreach (var function in Deployments.Functions)
         {
-            var updatedFunction = tasks.FirstOrDefault(t => t.Result.Deployment == function.Deployment);
-            updatedFunctions.Add(function with { Replicas = updatedFunction != null ? updatedFunction.Result.Replicas : function.Replicas });
+            var updatedFunction = tasks.Find(t => t.Result?.Deployment == function.Deployment);
```
Comment (g7ed6e) [inline src/SlimFaas/ReplicasService.cs:114]: In this foreach loop `.Result` Task property is accessed. The may cause deadlocks. We should have the following code before entering this loop.
```c#
ReplicaRequest?[] replicaResuests = await Task.WhenAll(tasks);
```
and loop through this array.

File: tests/SlimFaas.Tests/HistorySynchronizationWorkerShould.cs [hunk 4]
```diff
@@ -12,19 +13,17 @@ public async Task SyncLastTicksBetweenDatabaseAndMemory()
         var redisMockService = new RedisMockService();
         var historyHttpRedisService = new HistoryHttpRedisService(redisMockService);
         var kubernetesService = new Mock<IKubernetesService>();
-        var deploymentsInformations = new DeploymentsInformations()
+        var deploymentsInformations = new DeploymentsInformations(Functions: new List<DeploymentInformation>()
```
Comment (g7ed6e) [inline tests/SlimFaas.Tests/HistorySynchronizationWorkerShould.cs:16]: it looks weird to pass an empty list here, i think DeploymentsInformations should:
- be renamed DeploymentInformation
- initialize its own porperties internally

File: tests/SlimFaas.Tests/HistorySynchronizationWorkerShould.cs [hunk 5]
```diff
@@ -39,10 +38,10 @@ public async Task SyncLastTicksBetweenDatabaseAndMemory()
         await Task.Delay(200);
```
Comment (g7ed6e) [inline tests/SlimFaas.Tests/HistorySynchronizationWorkerShould.cs:39]: there's a lot of magic values that should either be centralized somewhere (a static class with const) or moved as const of the type they belong to.

File: tests/SlimFaas.Tests/ReplicasSynchronizationWorkerShould.cs [hunk 6]
```diff
@@ -1,22 +1,24 @@
 ﻿using System.Collections;
 using Microsoft.Extensions.Logging;
 using Moq;
+using SlimFaas.Kubernetes;
 
 namespace SlimFaas.Tests;
 
 public class DeploymentsTestData:IEnumerable<object[]>
```
Comment (g7ed6e) [inline tests/SlimFaas.Tests/ReplicasSynchronizationWorkerShould.cs:8]: another option to get strongly types is to inherit from TheoryData<T>. Then in the constructor, just call Add() to add tests iterations. This allow to remove the noise of object[].

**Pull request merged**